### PR TITLE
Align gradients API with TF (dl.grad returns a function f to be evaluated later)

### LINKED ISSUE #654: Change grad API
to allow `grad(f, x)(x0)`

### PR DESCRIPTION
**Pull request by dsmilkov**
Fixes #654 

This change is [Reviewable](https://reviewable.io/reviews/pair-code/deeplearnjs/690)

### CODE REVIEW

**Comment by nsthorat** [comment]
:lgtm_strong:

---

Reviewed 14 of 19 files at r1, 4 of 5 files at r2, 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[src/math/softmax.ts, line 108 at r3](https://reviewable.io:443/reviews/pair-code/deeplearnjs/690#-L56W-FCz1p6APndWOks:-L56W-FCz1p6APndWOkt:b-gz9n0u) ([raw file](https://github.com/pair-code/deeplearnjs/blob/fd13c81392b47ac82be88bf8ff4cc2713cb74e37/src/math/softmax.ts#L108)):*
> ```XML
>     }
>     // Use a custom gradient for numerical stability.
>     const customOp = customGrad(logits => {
> ```

this should take logits and label - the gradient should return both (this is supported in TF: https://www.tensorflow.org/api_docs/python/tf/nn/softmax_cross_entropy_with_logits_v2)

---

*[src/math/backends/gradients.ts, line 18 at r3](https://reviewable.io:443/reviews/pair-code/deeplearnjs/690#-L56XU_hSSxRN4nA5-sk:-L56XU_hSSxRN4nA5-sl:bzh870d) ([raw file](https://github.com/pair-code/deeplearnjs/blob/fd13c81392b47ac82be88bf8ff4cc2713cb74e37/src/math/backends/gradients.ts#L18)):*
> ```XML
>  */
> 
> import {ENV} from '../../environment';
> ```

What do you think about moving this file up a directory into math/? This seems like the wrong place for it, backends/ should be stuff for kernel implementation, not things that are backend-agnostic (my fault).

When you move it up, maybe you can move the tests in math_tests about gradients into gradients_test.ts?

This may also apply to tracking.ts, and even backend_engine.

---

*[src/math/backends/gradients.ts, line 49 at r3](https://reviewable.io:443/reviews/pair-code/deeplearnjs/690#-L56a91aOwbcBzi1KEjC:-L56XnSz8G1ImIoRQaHc:b-z4n6tk) ([raw file](https://github.com/pair-code/deeplearnjs/blob/fd13c81392b47ac82be88bf8ff4cc2713cb74e37/src/math/backends/gradients.ts#L49)):*
> ```XML
>    * `f(x)` must take a single tensor `x`. Returns another function `g(x, dy?)`,
>    * which when called gives `df/dx`. If `dy` is provided, the gradient of
>    * `dl.dot(f(x), dy)` w.r.t. `x` is computed instead.
> ```

f(x).mul(dy).sum()

---

*[src/math/backends/gradients.ts, line 98 at r3](https://reviewable.io:443/reviews/pair-code/deeplearnjs/690#-L56YSxxkc7Ulx5WI6wN:-L56YSxxkc7Ulx5WI6wO:b-22ze25) ([raw file](https://github.com/pair-code/deeplearnjs/blob/fd13c81392b47ac82be88bf8ff4cc2713cb74e37/src/math/backends/gradients.ts#L98)):*
> ```XML
>    */
>   @doc({heading: 'Training', subheading: 'Gradients'})
>   static grads<O extends Tensor>(f: (...args: Tensor[]) => O):
> ```

move this up next to grad

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pair-code/deeplearnjs/690#-:-L56alr3SQuP0N42XWF3:b-jbsf8z)*

**Comment by dsmilkov** [comment]
Reviewed 19 of 19 files at r1, 5 of 5 files at r2.
Review status: all files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---

*[src/math/softmax.ts, line 108 at r3](https://reviewable.io:443/reviews/pair-code/deeplearnjs/690#-L56W-FCz1p6APndWOks:-L56aw4dSSNBuzgUX74g:b-896fix) ([raw file](https://github.com/pair-code/deeplearnjs/blob/fd13c81392b47ac82be88bf8ff4cc2713cb74e37/src/math/softmax.ts#L108)):*
<details><summary><i>Previously, nsthorat (Nikhil Thorat) wrote…</i></summary><blockquote>

this should take logits and label - the gradient should return both (this is supported in TF: https://www.tensorflow.org/api_docs/python/tf/nn/softmax_cross_entropy_with_logits_v2)
</blockquote></details>

Done.

---

*[src/math/backends/gradients.ts, line 18 at r3](https://reviewable.io:443/reviews/pair-code/deeplearnjs/690#-L56XU_hSSxRN4nA5-sk:-L56b-EFyBXurXuIq-QY:bo6gtor) ([raw file](https://github.com/pair-code/deeplearnjs/blob/fd13c81392b47ac82be88bf8ff4cc2713cb74e37/src/math/backends/gradients.ts#L18)):*
<details><summary><i>Previously, nsthorat (Nikhil Thorat) wrote…</i></summary><blockquote>

What do you think about moving this file up a directory into math/? This seems like the wrong place for it, backends/ should be stuff for kernel implementation, not things that are backend-agnostic (my fault).

When you move it up, maybe you can move the tests in math_tests about gradients into gradients_test.ts?

This may also apply to tracking.ts, and even backend_engine.
</blockquote></details>

Yes, I'm planing on moving bunch of files out of `backends` folder (tracking, gradients, rename math_tests to engine_tests). Let's do that in a separate PR.

---

*[src/math/backends/gradients.ts, line 49 at r3](https://reviewable.io:443/reviews/pair-code/deeplearnjs/690#-L56a91aOwbcBzi1KEjC:-L56bViQZ5eb6VUIZWfS:b-896fix) ([raw file](https://github.com/pair-code/deeplearnjs/blob/fd13c81392b47ac82be88bf8ff4cc2713cb74e37/src/math/backends/gradients.ts#L49)):*
<details><summary><i>Previously, nsthorat (Nikhil Thorat) wrote…</i></summary><blockquote>

f(x).mul(dy).sum()
</blockquote></details>

Done.

---

*[src/math/backends/gradients.ts, line 98 at r3](https://reviewable.io:443/reviews/pair-code/deeplearnjs/690#-L56YSxxkc7Ulx5WI6wN:-L56bNsgJGVBrhyc1zBU:b-896fix) ([raw file](https://github.com/pair-code/deeplearnjs/blob/fd13c81392b47ac82be88bf8ff4cc2713cb74e37/src/math/backends/gradients.ts#L98)):*
<details><summary><i>Previously, nsthorat (Nikhil Thorat) wrote…</i></summary><blockquote>

move this up next to grad
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pair-code/deeplearnjs/690)*
